// @angular/common/esm2015/src/common_module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { NgModule } from '@angular/core';
import { COMMON_DIRECTIVES } from './directives/index';
import { DEPRECATED_PLURAL_FN, NgLocaleLocalization, NgLocalization, getPluralCase } from './i18n/localization';
import { COMMON_DEPRECATED_I18N_PIPES } from './pipes/deprecated/index';
import { COMMON_PIPES } from './pipes/index';
// Note: This does not contain the location providers,
// as they need some platform specific implementations to work.
/**
 * Exports all the basic Angular directives and pipes,
 * such as `NgIf`, `NgForOf`, `DecimalPipe`, and so on.
 * Re-exported by `BrowserModule`, which is included automatically in the root
 * `AppModule` when you create a new app with the CLI `new` command.
 *
 * * The `providers` options configure the NgModule's injector to provide
 * localization dependencies to members.
 * * The `exports` options make the declared directives and pipes available for import
 * by other NgModules.
 *
 * \@publicApi
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './directives/ng_class';
import * as ɵngcc2 from './directives/ng_component_outlet';
import * as ɵngcc3 from './directives/ng_for_of';
import * as ɵngcc4 from './directives/ng_if';
import * as ɵngcc5 from './directives/ng_template_outlet';
import * as ɵngcc6 from './directives/ng_style';
import * as ɵngcc7 from './directives/ng_switch';
import * as ɵngcc8 from './directives/ng_plural';
import * as ɵngcc9 from './pipes/async_pipe';
import * as ɵngcc10 from './pipes/case_conversion_pipes';
import * as ɵngcc11 from './pipes/json_pipe';
import * as ɵngcc12 from './pipes/slice_pipe';
import * as ɵngcc13 from './pipes/number_pipe';
import * as ɵngcc14 from './pipes/date_pipe';
import * as ɵngcc15 from './pipes/i18n_plural_pipe';
import * as ɵngcc16 from './pipes/i18n_select_pipe';
import * as ɵngcc17 from './pipes/keyvalue_pipe';
import * as ɵngcc18 from './pipes/deprecated/number_pipe';
import * as ɵngcc19 from './pipes/deprecated/date_pipe';
export class CommonModule {
}
CommonModule.ngModuleDef = ɵngcc0.ɵɵdefineNgModule({ type: CommonModule });
CommonModule.ngInjectorDef = ɵngcc0.ɵɵdefineInjector({ factory: function CommonModule_Factory(t) { return new (t || CommonModule)(); }, providers: [
        { provide: NgLocalization, useClass: NgLocaleLocalization },
    ] });
/*@__PURE__*/ ɵngcc0.ɵɵsetNgModuleScope(CommonModule, { declarations: [ɵngcc1.NgClass, ɵngcc2.NgComponentOutlet, ɵngcc3.NgForOf, ɵngcc4.NgIf, ɵngcc5.NgTemplateOutlet, ɵngcc6.NgStyle, ɵngcc7.NgSwitch, ɵngcc7.NgSwitchCase, ɵngcc7.NgSwitchDefault, ɵngcc8.NgPlural, ɵngcc8.NgPluralCase, ɵngcc9.AsyncPipe, ɵngcc10.UpperCasePipe, ɵngcc10.LowerCasePipe, ɵngcc11.JsonPipe, ɵngcc12.SlicePipe, ɵngcc13.DecimalPipe, ɵngcc13.PercentPipe, ɵngcc10.TitleCasePipe, ɵngcc13.CurrencyPipe, ɵngcc14.DatePipe, ɵngcc15.I18nPluralPipe, ɵngcc16.I18nSelectPipe, ɵngcc17.KeyValuePipe], exports: [ɵngcc1.NgClass, ɵngcc2.NgComponentOutlet, ɵngcc3.NgForOf, ɵngcc4.NgIf, ɵngcc5.NgTemplateOutlet, ɵngcc6.NgStyle, ɵngcc7.NgSwitch, ɵngcc7.NgSwitchCase, ɵngcc7.NgSwitchDefault, ɵngcc8.NgPlural, ɵngcc8.NgPluralCase, ɵngcc9.AsyncPipe, ɵngcc10.UpperCasePipe, ɵngcc10.LowerCasePipe, ɵngcc11.JsonPipe, ɵngcc12.SlicePipe, ɵngcc13.DecimalPipe, ɵngcc13.PercentPipe, ɵngcc10.TitleCasePipe, ɵngcc13.CurrencyPipe, ɵngcc14.DatePipe, ɵngcc15.I18nPluralPipe, ɵngcc16.I18nSelectPipe, ɵngcc17.KeyValuePipe] });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(CommonModule, [{
        type: NgModule,
        args: [{
                declarations: [COMMON_DIRECTIVES, COMMON_PIPES],
                exports: [COMMON_DIRECTIVES, COMMON_PIPES],
                providers: [
                    { provide: NgLocalization, useClass: NgLocaleLocalization },
                ]
            }]
    }], null, null);
const ɵ0 = getPluralCase;
/**
 * A module that contains the deprecated i18n pipes.
 *
 * @deprecated from v5
 * \@publicApi
 */
export class DeprecatedI18NPipesModule {
}
DeprecatedI18NPipesModule.ngModuleDef = ɵngcc0.ɵɵdefineNgModule({ type: DeprecatedI18NPipesModule });
DeprecatedI18NPipesModule.ngInjectorDef = ɵngcc0.ɵɵdefineInjector({ factory: function DeprecatedI18NPipesModule_Factory(t) { return new (t || DeprecatedI18NPipesModule)(); }, providers: [{ provide: DEPRECATED_PLURAL_FN, useValue: ɵ0 }] });
/*@__PURE__*/ ɵngcc0.ɵɵsetNgModuleScope(DeprecatedI18NPipesModule, { declarations: [ɵngcc18.DeprecatedDecimalPipe, ɵngcc18.DeprecatedPercentPipe, ɵngcc18.DeprecatedCurrencyPipe, ɵngcc19.DeprecatedDatePipe], exports: [ɵngcc18.DeprecatedDecimalPipe, ɵngcc18.DeprecatedPercentPipe, ɵngcc18.DeprecatedCurrencyPipe, ɵngcc19.DeprecatedDatePipe] });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(DeprecatedI18NPipesModule, [{
        type: NgModule,
        args: [{
                declarations: [COMMON_DEPRECATED_I18N_PIPES],
                exports: [COMMON_DEPRECATED_I18N_PIPES],
                providers: [{ provide: DEPRECATED_PLURAL_FN, useValue: ɵ0 }]
            }]
    }], null, null);
export { ɵ0 };
